File: src/worldgen.ts
import { Chunk, GrassBlock } from "./world"
import { Noise } from 'noisejs';

export interface ChunkGenerator {
    generateChunk(cx: number, cy: number): Chunk;
    seed(seed: number): void;
}

export class PerlinChunkGenerator implements ChunkGenerator {
    private noise: Noise;
    
    constructor(seed?: number) {
        this.noise = new Noise();
        if (seed != null) {
            this.seed(seed);
        }
    }

    private generateHeight(cx: number, cy: number): number[] {
        const heights = new Array<number>(Chunk.SizeX*Chunk.SizeY);
        for (let j=0; j<Chunk.SizeY; j++) {
            for (let i=0; i<Chunk.SizeX; i++) {
                const noise = this.noise.perlin2(cx + i / Chunk.SizeX,
                                                 cy + j / Chunk.SizeY);
                heights[Chunk.SizeX*j+i] = Math.round((noise + 1.0) / 2.0 * Chunk.SizeZ);
            }
        }
        return heights;
    }
    
    generateChunk(cx: number, cy: number): Chunk {
        const heightMap = this.generateHeight(cx, cy);
        const chunk = new Chunk(cx, cy);
        for (let j=0; j<Chunk.SizeY; j++) {
            for (let i=0; i<Chunk.SizeX; i++) {
                const k = heightMap[Chunk.SizeX*j+i];
                chunk.blocks[Chunk.index(i, j, k)] = GrassBlock;
            }
        }
        return chunk;
    }

    seed(seed: number) {
        this.noise.seed(seed);
    }
}

File: client2/src/worldgen.cpp
#include "./worldgen.h"
#include "../PerlinNoise/PerlinNoise.hpp"

using namespace tinaxcraft;

PerlinNoiseWorldGen::PerlinNoiseWorldGen(int32_t seed) : seed_(seed), noise_(seed)
{
}

Chunk *PerlinNoiseWorldGen::generateChunk(const ChunkCoord &coord)
{
    const auto &heightMap = generateHeight(coord);

    auto *c = new Chunk(coord);

    for (int32_t x = 0; x < Chunk::SizeX; x++)
    {
        for (int32_t z = 0; z < Chunk::SizeZ; z++)
        {
            const auto &height = heightMap.at(x * Chunk::SizeZ + z);
            for (int32_t y = 0; y <= height; y++)
            {
                // set to dirt block
                c->setBlock(CoordInChunk(x, y, z), Block(1));
            }
        }
    }

    return c;
}

std::vector<int32_t> PerlinNoiseWorldGen::generateHeight(const ChunkCoord &coord)
{
    const auto SizeX = Chunk::SizeX;
    const auto SizeZ = Chunk::SizeZ;
    std::vector<int32_t> heights(SizeX * SizeZ);

    for (int i = 0; i < SizeX; i++)
    {
        for (int j = 0; j < SizeZ; j++)
        {
            const auto noise = noise_.noise2D_01(coord.cx + ((float)i / SizeX), coord.cz + ((float)j / SizeZ));
            heights.at(i * SizeZ + j) = 64 + static_cast<int32_t>(noise * 16.0);
        }
    }

    return heights;
}